<a href="https://colab.research.google.com/github/xinyang4O4/MAT421/blob/main/ModuleC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 19.1 Root Finding Problem Statement

The root (zero) of a function, f(x), is an $x_r$ such that $f(x_r)$ = 0.

Finding root for a function f sometimes could be hard, so instead we can find numerical approximation.

In [ ]:
#solve for a root of f(x) = cos(x)-x near 1
import numpy as np
from scipy import optimize

f = lambda x: np.cos(x)-x
r = optimize.fsolve(f, 1)
print("r =", r)

# Verify the solution is a root
result = f(r) #involves a floating point division, so the answer will be 0. which is the same as 0.0
print("result=", result)

r = [0.73908513]
result= [0.]


In [ ]:
#solve for a root of f(x) = 1/x which does not exitst
f = lambda x: 1/x

r, infodict, ier, mesg = optimize.fsolve(f, 3, full_output=True)
print("r =", r)

result = f(r)
print("result=", result)

print(mesg) #This message will return that no root is found

r = [5.28071045e+83]
result= [1.89368459e-84]
The number of calls to function has reached maxfev = 400.


# 19.2 Tolerance

Tolerance is the level of error that is acceptable for an engineering application, and when the comuter finds a solution that has error smaller than the tolerance, we say that it has converged to a solution.

In [ ]:
tol = 1e-10
f = lambda x: x**2 +tol/2

#calculates the error when input error is less than tol
r, infodict, ier, mesg = optimize.fsolve(f, 3, full_output=True)
error = f(r)
print("error = ", error)

yzero = f(0)
print("f(0) = ", yzero)

error= [5.03815109e-11]
f(0) 5e-11


In [ ]:
tol = 1e-10
f = lambda x: x**2+tol/2

#calculates the error when input error is less than tol
xi = -tol/4
xj = tol/4
e = tol/2
result = (f(xj)-f(xi))/2
print("result = ", result)
print("e = ", e)

result =  0.0
e =  5e-11


# 19.3 Bisection Method

The intermediate Value Theorem says that for a continuous function f such that $sign(f(a)) \neq sign(f(b))$ there must exists c between a and b and f(c) = 0.

The bisection method uses the intermediate value theorem iteratively to find roots. It basiclly determains the zero by letting m recursively close to 0 such that f(m) < tolerance

In [ ]:
import numpy as np

def my_bisection(f, a, b, tol):
    # approximates a root, R, of f bounded
    # by a and b to within tolerance
    # | f(m) | < tol with m the midpoint
    # between a and b Recursive implementation

    # check if a and b bound a root
    if np.sign(f(a)) == np.sign(f(b)):
        raise Exception(
         "The scalars a and b do not bound a root")

    # get midpoint
    m = (a + b)/2

    if np.abs(f(m)) < tol:
        # stopping condition, report m as root
        return m
    elif np.sign(f(a)) == np.sign(f(m)):
        # case where m is an improvement on a.
        # Make recursive call with a = m
        return my_bisection(f, m, b, tol)
    elif np.sign(f(b)) == np.sign(f(m)):
        # case where m is an improvement on b.
        # Make recursive call with b = m
        return my_bisection(f, a, m, tol)

In [ ]:
f = lambda x: x**2 - 2

r1 = my_bisection(f, 0, 2, 0.1)
print("r1 =", r1)
r01 = my_bisection(f, 0, 2, 0.01)
print("r01 =", r01)

print("f(r1) =", f(r1))
print("f(r01) =", f(r01))

r1 = 1.4375
r01 = 1.4140625
f(r1) = 0.06640625
f(r01) = -0.00042724609375


We can tell that the less the tolerance is, the closer the answer is to zero. (It will only work if a and b bound a root)

# 19.4 Newton-Raphson Method

The Newton-Raphson method is another way to find the root. We first randomly pick a $x_0$, then determain the approximation $x_n=x_{n-1}−\frac{f(x_{n-1})}{f'(x_{n-1})}$

The case that this method won't work so well is that when the derivative of the result is complicate. Another case would be for the choice of the approximation.

In [ ]:
def my_newton(f, df, x0, tol):
    # output is an estimation of the root of f
    # using the Newton Raphson method
    # recursive implementation
    if abs(f(x0)) < tol:
        return x0
    else:
        return my_newton(f, df, x0 - f(x0)/df(x0), tol)


In [ ]:
import numpy as np

f = lambda x: x**2 - 4
f_prime = lambda x: 2*x

estimate = my_newton(f, f_prime, 5, 1e-6)
print("estimate =", estimate)
print("result =", 2)

estimate = 2.000000000006711
result = 2


The result is really close to the real root and it's faster to compile than the Bisection Method

# 19.5 Root Finding in Python

In python, we can always use the f_solve method from the scipy.optimize directly.

In [ ]:
from scipy.optimize import fsolve

f = lambda x: 5*x**4+x**3-6*x**2-3*x+2
fsolve(f, [1,100])

array([1.09232514, 1.09338332])

Then we know that the two roots are 1.09232514 and 1.09338332